homeassistant/components/ssdp/__init__.py

```diff
@@ -567,34 +559,44 @@ async def _async_headers_to_discovery_info(
 
         location = headers["location"]
         info_desc = await self._async_get_description_dict(location)
-        return discovery_info_from_headers_and_description(headers, info_desc)
+        return discovery_info_from_headers_and_description(
+            ssdp_device, headers, info_desc
+        )
 
     async def async_get_discovery_info_by_udn_st(
         self, udn: str, st: str
     ) -> SsdpServiceInfo | None:
         """Return discovery_info for a udn and st."""
-        if headers := self._all_headers_from_ssdp_devices.get((udn, st)):
-            return await self._async_headers_to_discovery_info(headers)
+        for ssdp_device in self._ssdp_devices:
+            if ssdp_device.udn == udn:
+                if headers := ssdp_device.combined_headers(st):
+                    return await self._async_headers_to_discovery_info(
+                        ssdp_device, headers
+                    )
         return None
 
     async def async_get_discovery_info_by_st(self, st: str) -> list[SsdpServiceInfo]:
         """Return matching discovery_infos for a st."""
         return [
-            await self._async_headers_to_discovery_info(headers)
-            for udn_st, headers in self._all_headers_from_ssdp_devices.items()
-            if udn_st[1] == st
+            await self._async_headers_to_discovery_info(
+                ssdp_device, ssdp_device.combined_headers(st)
```

Comment:
Note the double `ssdp_device.combined_headers(st)` here (and in the `if`).

Reply:
yikes. thats an expensive function (probably the most expensive one actually)

Maybe
```diff
diff --git a/homeassistant/components/ssdp/__init__.py b/homeassistant/components/ssdp/__init__.py
index 1aca128f136..e4db2ac9df1 100644
--- a/homeassistant/components/ssdp/__init__.py
+++ b/homeassistant/components/ssdp/__init__.py
@@ -579,10 +579,10 @@ class Scanner:
         """Return matching discovery_infos for a st."""
         return [
             await self._async_headers_to_discovery_info(
-                ssdp_device, ssdp_device.combined_headers(st)
+                ssdp_device, headers
             )
             for ssdp_device in self._ssdp_devices
-            if ssdp_device.combined_headers(st)
+            if (headers := ssdp_device.combined_headers(st))
         ]
 
     async def async_get_discovery_info_by_udn(self, udn: str) -> list[SsdpServiceInfo]:
```

Reply:
Thanks, was trying without the parenthesis, but that gave me syntax errors. Learned something today!